### PR TITLE
Rework of DTLS code, add DTLS PKI/RPK

### LINKED ISSUE #4: Public Key Infrastructure/Raw Public Key Support
## Summary

Allow for usage of Raw Public Keys and/or a Public Key Infrastructure instead of pre-shared keys to establish DTLS sessions.

## Description

We currently do not expose the API required for using raw public keys or certificates instead of symmetric pre-shared keys for DTLS session establishment. 
This should obviously be changed.

## Implementation steps

- [ ] Design Rust-idiomatic API
- [ ] Implement wrapper code
- [ ] Add tests (ideally also for PSK, since PSK currently has no tests)

(Issue moved over from GitLab)

### LINKED ISSUE #30: Enable DTLS EC JPAKE support
libcoap 4.3.5 allows using EC JPAKE ([RFC 8236](https://datatracker.ietf.org/doc/html/rfc8236#section-3)) for key exchanges.

We should support it too.

### PR DESCRIPTION
**Pull request by pulsastrix**
This PR refactors the cryptography related code to improve usage ergonomics and to add DTLS support using PKI/RPK.

~~Still a WIP, not ready for reviews yet.~~

Closes #30, Closes #4.

### CODE REVIEW

**Comment by pulsastrix** [comment]
This should now be ready for a review (I will apply the clippy warnings alongside other review comments).

(Side note: apparently, the CI pipeline still needs some improvements, as the clippy warnings should definitely fail the pipeline. Will fix this in a new merge request alongside some other pipeline improvements.)

**Review by JKRhb — COMMENTED** [review]
See a first couple of nits below :)

File: libcoap/src/crypto/mod.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+//! - (D)TLS using pre-shared keys (PSK): The easiest method, uses a symmetric/pre-shared key to
+//!   perform authentication (see [RFC 4279](https://datatracker.ietf.org/doc/html/rfc4279)).
+//! - (D)TLS using raw public keys (RPK): Uses asymmetric key pairs for authentication. The peer's
+//!   public key must be known in advance and must be validated by the library user.
+//! - (D)TLS using a public key infrastructure (PKI): Uses asymmetric key pairs signed by a
+//!   certificate authority, which are authenticated by the TLS library using a set of
+//!   pre-configured (or provided) root certificate authorities (the way most of the internet works).
+//! - OSCORE (*unsupported by libcoap-rs, see
+//!   [issue #23](https://github.com/namib-project/libcoap-rs/issues/23)*): Uses Object Security for
+//!   Constrained RESTful Environments (OSCORE, RFC 8613) to encrypt messages on the application
```
[the remark this answers is not in the document]
Comment (pulsastrix) [reply libcoap/src/crypto/mod.rs:22]:
```suggestion
//!   Constrained RESTful Environments (OSCORE, [RFC 8613](https://datatracker.ietf.org/doc/html/rfc8613)) to encrypt messages on the application
```

File: libcoap/src/crypto/pki_rpk/mod.rs [hunk 2]
```diff
@@ -0,0 +1,888 @@
...
+            .finish()
+    }
+}
+
+impl<KTY: KeyType> Drop for PkiRpkContextInner<'_, KTY> {
+    fn drop(&mut self) {
+        for key_ref in std::mem::take(&mut self.provided_key_descriptors).into_iter() {
+            // SAFETY: If the inner context is dropped, this implies that the pointers returned in
+            // previous callbacks are no longer used (because of the contracts of apply_to_context()
+            // and create_raw_session()). We can therefore restore and drop these values without
```
[the remark this answers is not in the document]
Comment (pulsastrix) [reply libcoap/src/crypto/pki_rpk/mod.rs:666]: Not necessarily, as these aren't docstrings and therefore not interpreted as markdown.

File: libcoap/src/crypto/pki_rpk/pki.rs [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            None,
+            private_key_type,
+        )
+    }
+}
+
+impl PkiKeyDef<Pkcs11KeyComponent, Pkcs11KeyComponent, Pkcs11KeyComponent> {
+    /// Creates a new key definition using PKCS11 URIs as components.
+    ///
+    /// See the documentation of [PkiKeyDef::new] for more information on the parameters, especially
```
Comment (pulsastrix) [inline libcoap/src/crypto/pki_rpk/pki.rs:287]:
```suggestion
    /// See the documentation of [`PkiKeyDef::new`] for more information on the parameters, especially
```

File: libcoap/src/crypto/pki_rpk/rpk.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    ///
+    /// `asn1_encoded_key` should be the certificate structure defined in
+    /// [RFC 7250, section 3](https://datatracker.ietf.org/doc/html/rfc7250#section-3), which you
+    /// might be able to parse with crates like
+    /// [x509-cert](https://docs.rs/x509-cert/latest/x509_cert/index.html) and
+    /// [spki](https://docs.rs/spki/0.7.3/spki/index.html) to obtain and match the
+    /// SubjectPublicKeyInformation encoded within.
+    ///
+    /// See [the libcoap documentation](https://libcoap.net/doc/reference/4.3.5/group__dtls.html#gaef7a2800757a4922102311c94c3fa529)
+    /// for more background information
```
Comment (pulsastrix) [inline libcoap/src/crypto/pki_rpk/rpk.rs:84]:
```suggestion
    /// for more information.
```

File: libcoap/src/crypto/psk/client.rs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+#[derive(Debug)]
+struct ClientPskContextInner<'a> {
+    /// Raw configuration object.
+    raw_cfg: Box<coap_dtls_cpsk_t>,
+    /// User-supplied key provider.
+    key_provider: Option<Box<dyn ClientPskHintKeyProvider<'a> + 'a>>,
+    /// Store for `coap_dtls_cpsk_info_t` instances that we provided in previous identity hint
+    /// callback invocations.
+    ///
+    /// The stored pointers *must* all be created from Box::into_raw().
```
[the remark this answers is not in the document]
Comment (pulsastrix) [reply libcoap/src/crypto/psk/client.rs:252]:
```suggestion
    /// The stored pointers *must* all be created from [`Box::into_raw`].
```

File: libcoap/src/session/server.rs [hunk 6]
```diff
@@ -102,6 +102,8 @@ impl CoapServerSession<'_> {
     ///
     /// # Safety
     /// The provided pointer must be valid for the entire lifetime of this struct.
+    /// The provided session's app data must be a valid argument to
+    /// `CoapFfiRawCell<CoapServerSessionInner>::clone_raw_rc`.
```
Comment (JKRhb) [inline libcoap/src/session/server.rs:106]: Should this also be a link instead?

File: libcoap/src/session/mod.rs [hunk 7]
```diff
@@ -443,7 +443,9 @@ impl<'a> CoapSession<'a> {
     /// Panics if the given pointer is a null pointer.
     ///
     /// # Safety
-    /// The provided pointer must be valid.
+    /// The provided pointer must be valid, the provided session's app data must be a valid argument
+    /// to `CoapFfiRawCell<CoapClientSessionInner>::clone_raw_rc` or
+    /// `CoapFfiRawCell<CoapServerSessionInner>::clone_raw_rc` (depending on the session type).
```
Comment (JKRhb) [inline libcoap/src/session/mod.rs:448]: Are there square brackets missing?

```suggestion
    /// to [`CoapFfiRawCell<CoapClientSessionInner>::clone_raw_rc`] or
    /// [`CoapFfiRawCell<CoapServerSessionInner>::clone_raw_rc`] (depending on the session type).
```

File: libcoap/src/context.rs [hunk 8]
```diff
@@ -286,18 +383,18 @@ impl CoapContext<'_> {
...
     }
 
     /// TODO
-    #[cfg(all(feature = "tcp", feature = "dtls"))]
+    /*#[cfg(all(feature = "tcp", dtls))]
     pub fn add_endpoint_tls(&mut self, _addr: SocketAddr) -> Result<(), EndpointCreationError> {
         todo!()
         // TODO: self.add_endpoint(addr, coap_proto_t::COAP_PROTO_TLS)
-    }
+    }*/
```
Comment (JKRhb) [inline libcoap/src/context.rs:397]: Hmm, would it be better to only use `//` for commenting this out here?

**Review by JKRhb — APPROVED** [review]
See a last couple of nits below, otherwise LGTM! :)

File: libcoap-sys/src/lib.rs [hunk 9]
```diff
@@ -200,27 +200,27 @@ pub fn coap_startup_with_feature_checks() {
         #[cfg(feature = "dtls-cid")]
         // SAFETY: Function is always safe to call.
         if unsafe { coap_dtls_cid_is_supported() != 1 } {
-            panic!("Required feature \"dtls\" is not supported by libcoap")
+            panic!("Required feature \"dtls-cid\" is not supported by libcoap")
```
[the remark this answers is not in the document]
Comment (pulsastrix) [reply libcoap-sys/src/lib.rs:203]: Converted to issue in #34

**Pull request merged**